ktor-client/ktor-client-features/ktor-client-tracing/common/src/TracingWrapper.kt

```diff
@@ -0,0 +1,181 @@
+ */
+class TracingWrapper<T : HttpClientEngineConfig>(
+    private val delegate: HttpClientEngineFactory<T>,
+    private val tracer: Tracer
+) : HttpClientEngineFactory<T> {
+    override fun create(block: T.() -> Unit): HttpClientEngine {
+        val engine = delegate.create(block)
+        return EngineWithTracer(engine, tracer)
+    }
+}
+
+/**
+ * Engine with tracer that wraps request execution into tracing functionality. Calls correspondent [tracer] method on
+ * every important processing event.
+ */
+internal class EngineWithTracer(private val delegate: HttpClientEngine, private val tracer: Tracer) :
+    HttpClientEngine by delegate {
+
+    private val sequence = atomic(0)
+
+    @InternalAPI
+    override suspend fun execute(data: HttpRequestData): HttpResponseData {
+        val requestId = "${sequence.getAndIncrement()}"
+        val isWebSocket = data.body is ClientUpgradeContent
+        if (isWebSocket) {
+            tracer.webSocketCreated(requestId, data.url.toString())
+            tracer.webSocketWillSendHandshakeRequest(requestId, data)
+        } else {
+            tracer.requestWillBeSent(requestId, data)
+        }
+
+        try {
+            val result = delegate.execute(data)
+            if (isWebSocket) {
+                tracer.webSocketHandshakeResponseReceived(requestId, data, result)
+            } else {
+                tracer.responseHeadersReceived(requestId, data, result)
+            }
+
+            coroutineContext[Job]!!.invokeOnCompletion {
+                if (isWebSocket) {
+                    tracer.webSocketClosed(requestId)
+                } else {
+                    tracer.responseReadFinished(requestId)
+                }
+            }
+
+            return with(result) {
+                HttpResponseData(
+                    statusCode,
+                    requestTime,
+                    headers,
+                    version,
+                    if (isWebSocket) {
+                        WebSocketSessionTracer(
+                            requestId,
+                            tracer,
+                            result.body as DefaultWebSocketSession
+                        )
+                    } else {
+                        tracer.interpretResponse(
+                            requestId,
+                            headers[HttpHeaders.ContentType],
+                            headers[HttpHeaders.ContentEncoding],
+                            result.body
+                        )!!
+                    },
+                    callContext
+                )
+            }
+        } catch (cause: Throwable) {
+            tracer.httpExchangeFailed(requestId, cause.message!!)
+            throw cause
+        }
+    }
+
+    @InternalAPI
+    override fun install(client: HttpClient) {
+        super.install(client)
+    }
+}
+
+internal class WebSocketSessionTracer(requestId: String, tracer: Tracer, private val delegate: DefaultWebSocketSession) :
```

Comment:
Consider extracting this classes in separate files regarding their purpose: websocket, channel, etc

Reply:
Done.